src/data/models/LevelStats.js

```diff
@@ -0,0 +1,263 @@
+    type: Sequelize.INTEGER,
+    allowNull: true,
+  },
+  BattleTopBattleIndex: {
+    type: Sequelize.INTEGER,
+    allowNull: true,
+  },
+
+  // JSON array
+  LeaderHistory: {
+    type: Sequelize.TEXT('long'),
+    allowNull: true,
+    ...jsonAccessors('LeaderHistory', []),
+  },
+  LeaderCount: {
+    type: Sequelize.INTEGER,
+    allowNull: true,
+  },
+  UniqueLeaderCount: {
+    type: Sequelize.INTEGER,
+    allowNull: true,
+  },
+
+  KuskisF: {
+    type: Sequelize.INTEGER,
+    allowNull: true,
+  },
+  KuskiAll: {
+    type: Sequelize.INTEGER,
+    allowNull: true,
+  },
+  // need all IDs stored to build counts incrementally
+  KuskiIdsF: {
+    type: Sequelize.TEXT,
+    allowNull: true,
+    ...jsonAccessors('KuskiIdsF', []),
+  },
+  KuskiIdsAll: {
+    type: Sequelize.TEXT,
+    allowNull: true,
+    ...jsonAccessors('KuskiIdsAll', []),
+  },
+
+  // level stats should agree with the time table starting
+  // from TimeIndex 1 up until this point.
+  LastPossibleTimeIndex: {
+    type: Sequelize.INTEGER,
+    allowNull: false,
+    defaultValue: 0,
+  },
+
+  // virtual column returning an array of objects.
+  // it's much easier to work with top times like this and this
+  // should probably be passed to the front-end whenever possible.
+  TopXTimes: {
+    type: Sequelize.VIRTUAL,
+    get() {
+      return _.range(0, countTopXTimes).map(i => {
+        // must have same indexes as times table, or code elsewhere will break.
+        return {
+          KuskiIndex: this.getDataValue(`TopKuskiIndex${i}`),
+          Time: this.getDataValue(`TopTime${i}`),
+          TimeIndex: this.getDataValue(`TopTimeIndex${i}`),
+          Driven: this.getDataValue(`TopDriven${i}`),
+        };
+      });
+    },
+  },
+};
+
+// array of non virtual column names
+const cols = _.toPairs(ddl)
+  .filter(arr => arr[1].type !== Sequelize.VIRTUAL)
+  .map(arr => arr[0]);
+
+class LevelStats extends Model {
+  // map an array of ids to existing instances or null
+  // returns an object, indexed by ID.
+  static mapIds = async levelIds => {
+    const records = await LevelStats.findAll({
+      where: {
+        LevelIndex: levelIds,
+      },
+    });
+
+    return PlayStats.mapIdsToRecordsOrNull(levelIds, records, 'LevelIndex');
+  };
+
+  // useful for using .bulkCreate() to upsert
+  static getUpdateOnDuplicateKeys = () => {
+    return cols.filter(
+      key => !_.includes(['LevelStatsIndex', 'LevelIndex'], key),
+    );
+  };
+
+  /**
+   * Build the record to get inserted into the database.
+   *
+   * @param {Array<Object>} times - times that share the same level index
+   * @param {LevelStats|null} prev - existing LevelStats row
+   * @param LastPossibleTimeIndex
+   * @returns {Array} - updates (array of objects), and performance data
+   */
+  static buildUpdate = (times, prev, LastPossibleTimeIndex) => {
+    const track = getPerfTracker('buildUpdate');
+
+    assert(prev === null || _.isObjectLike(prev), `bad type: ${typeof prev}`);
```

Comment:
gonna change this useless check. if you pass in the next most logical thing, an empty object, then all goes to shit because the code everywhere checks !== null. Anyways its not a problem as is, the code works, but a useless attempt at protecting me from myself.